src/cpp/session/modules/rmarkdown/NotebookQueue.cpp

```diff
@@ -0,0 +1,560 @@
+      arr.push_back(chunkId);
+
+      // formulate request body
+      json::Object rpc;
+      rpc["method"] = "console_input";
+      rpc["params"] = arr;
+      rpc["clientId"] = clientEventService().clientId();
+
+      // serialize RPC body and send it to helper thread for submission
+      std::ostringstream oss;
+      json::write(rpc, oss);
+      pInput_->enque(oss.str());
+   }
+
+   Error executeNextUnit(ExpressionMode mode)
+   {
+      // no work to do if we have no documents
+      if (queue_.empty())
+         return Success();
+
+      // get the next execution unit from the current queue
+      boost::shared_ptr<NotebookDocQueue> docQueue = *queue_.begin();
+      if (docQueue->complete())
+         return Success();
+
+      boost::shared_ptr<NotebookQueueUnit> unit = docQueue->firstUnit();
+
+      // establish execution context for the unit
+      json::Object options;
+      Error error = unit->parseOptions(&options);
+      if (error)
+         LOG_ERROR(error);
+
+      // in batch mode, make sure unit should be evaluated -- note that
+      // eval=FALSE units generally do not get sent up in the first place, so
+      // if we're here it's because the unit has eval=<expr>
+      if (unit->execMode() == ExecModeBatch)
+      {
+         bool eval = true;
+         json::readObject(options, "eval", &eval);
+         if (!eval)
+         {
+            return skipUnit();
+         }
+      }
+
+      // compute context
+      std::string ctx = docQueue->commitMode() == ModeCommitted ?
+         kSavedCtx : notebookCtxId();
+
+      // compute engine
+      std::string engine = "r";
+      json::readObject(options, "engine", &engine);
+      if (engine == "r")
+      {
+         execContext_ = boost::make_shared<ChunkExecContext>(
+            unit->docId(), unit->chunkId(), ctx, unit->execScope(), options,
+            docQueue->pixelWidth(), docQueue->charWidth());
+         execContext_->connect();
+         execUnit_ = unit;
+         enqueueExecStateChanged(ChunkExecStarted, options);
+      }
+      else
+      {
+         // execute with alternate engine
+         std::string innerCode;
+         error = unit->innerCode(&innerCode);
+         if (error)
+         {
+            LOG_ERROR(error);
+         }
+         else
+         {
+            execUnit_ = unit;
+            enqueueExecStateChanged(ChunkExecStarted, options);
+            error = executeAlternateEngineChunk(
+               unit->docId(), unit->chunkId(), ctx, engine, innerCode, options);
+            if (error)
+               LOG_ERROR(error);
+         }
+      }
+
+      // if there was an error, skip the chunk
+      if (error)
+         return skipUnit();
+
+      if (engine == "r")
+      {
+         error = executeCurrentUnit(ExprModeNew);
+         if (error)
+            LOG_ERROR(error);
+      }
+
+      return Success();
+   }
+
+   // main function for thread which receives console input
+   void consoleThreadMain()
```

Comment:
Thanks, that's good to know! I'm inclined to leave it since it's performed well in testing (and I'm not sure if there could be other problems with issuing a blocking RPC from this thread), but I'll do a trial run at `schedulePeriodicWork` if you have any concerns about the thread-based approach.